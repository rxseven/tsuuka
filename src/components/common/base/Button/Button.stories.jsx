import React from 'react';
import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react';

import Button from './index';

const title = 'Components/Common/Base/Button';
const props = { onClick: action('onClick') };

storiesOf(title, module)
  .addParameters({
    info: {
      text: `
    **Usage:** for more information, see [Bootstrap - Buttons](https://getbootstrap.com/docs/4.1/components/buttons/).
  `
    }
  })
  .add('normal (default)', () => <Button {...props}>Normal Button</Button>)
  .add('look', () => (
    <Button {...props} look="primary">
      Primary
    </Button>
  ))
  .add('outline', () => (
    <Button {...props} outline="primary">
      Primary
    </Button>
  ))
  .add('size', () => (
    <Button {...props} look="primary" size="lg">
      Large
    </Button>
  ))
  .add('type', () => (
    <Button {...props} look="primary" type="submit">
      Submit
    </Button>
  ))
  .add('block', () => (
    <Button {...props} block look="primary">
      OK
    </Button>
  ))
  .add('disabled', () => (
    <Button {...props} isLoading look="primary">
      OK
    </Button>
  ))
  .add('with icon', () => (
    <Button {...props} icon="cloud-upload-alt" look="primary" />
  ))
  .add('with icon & text', () => (
    <Button {...props} icon="cloud-upload-alt" look="primary">
      Upload
    </Button>
  ))
  .add('with spinner', () => (
    <Button {...props} isLoading look="primary" spinner>
      Upload
    </Button>
  ))
  .add('passive', () => (
    <Button icon="cloud-upload-alt" look="primary" passive>
      Upload
    </Button>
  ));
