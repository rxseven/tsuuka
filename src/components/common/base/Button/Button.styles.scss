// Bootstrap Buttons
// https://getbootstrap.com/docs/4.1/components/buttons/

.btn {
  position: relative;

  &:not(:disabled):not(.disabled) {
    cursor: pointer;
  }
}
