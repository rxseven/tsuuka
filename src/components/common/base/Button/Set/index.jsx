import PropTypes from 'prop-types';
import exact from 'prop-types-exact';
import React from 'react';
import styled from 'styled-components';

const Frame = styled.div`
  align-items: center;
  display: flex;

  > :not(:last-child) {
    margin-right: 0.25rem;
  }

  > :not(:first-child) {
    margin-left: 0.25rem;
  }
`;

const propTypes = exact({
  children: PropTypes.node.isRequired
});

function Set({ children }) {
  return <Frame>{children}</Frame>;
}

Set.propTypes = propTypes;

export default Set;
