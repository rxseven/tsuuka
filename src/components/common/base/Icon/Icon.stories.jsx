import React from 'react';
import { storiesOf } from '@storybook/react';

import Icon from './index';

const title = 'Components/Common/Base/Icon';

storiesOf(title, module)
  .addParameters({
    info: {
      text: `
      **Usage:** for more information, see [@fortawesome/react-fontawesome](https://www.npmjs.com/package/@fortawesome/react-fontawesome).
    `
    }
  })
  .add('solid (default)', () => <Icon icon="home" />)
  .add('regular', () => <Icon icon={['far', 'trash-alt']} />)
  .add('brand', () => <Icon icon={['fab', 'github-alt']} />);
