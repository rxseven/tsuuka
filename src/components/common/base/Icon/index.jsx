import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

function Icon(props) {
  return <FontAwesomeIcon {...props} />;
}

export default Icon;
