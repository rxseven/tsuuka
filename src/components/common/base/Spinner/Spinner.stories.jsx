import React from 'react';
import { storiesOf } from '@storybook/react';

import Spinner from './index';

const title = 'Components/Common/Base/Spinner';

storiesOf(title, module)
  .add('default', () => <Spinner />)
  .add('custom color', () => <Spinner color="#007BFD" />)
  .add('custom size', () => <Spinner size={15} />)
  .add('mix', () => <Spinner color="#007BFD" size={30} />);
