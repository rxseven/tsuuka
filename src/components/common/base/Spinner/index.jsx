import PropTypes from 'prop-types';
import exact from 'prop-types-exact';
import React from 'react';
import { ThreeBounce } from 'better-react-spinkit';

const propTypes = exact({
  color: PropTypes.string,
  size: PropTypes.number
});

const defaultProps = {
  color: '#999',
  size: 7
};

const options = {
  duration: '1.25s',
  gutter: 5
};

function Spinner(props) {
  return <ThreeBounce {...options} {...props} />;
}

Spinner.propTypes = propTypes;
Spinner.defaultProps = defaultProps;

export default Spinner;
