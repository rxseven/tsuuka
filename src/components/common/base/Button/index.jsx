import cx from 'classnames';
import { omit } from 'lodash';
import PropTypes from 'prop-types';
import React from 'react';
import { isBrowser, isMobile } from 'react-device-detect';
import styled from 'styled-components';

import Icon from 'components/common/base/Icon';
import Spinner from 'components/common/base/Spinner';

import Set from './Set';
import './Button.styles.scss';

const propTypes = {
  block: PropTypes.bool,
  children: PropTypes.string,
  icon: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string)
  ]),
  isLoading: PropTypes.bool,
  look: PropTypes.string,
  minimal: PropTypes.bool,
  outline: PropTypes.string,
  passive: PropTypes.bool,
  size: PropTypes.string,
  spinner: PropTypes.bool,
  type: PropTypes.string
};

const defaultProps = {
  block: false,
  children: '',
  icon: '',
  isLoading: false,
  look: 'secondary',
  minimal: false,
  outline: '',
  passive: false,
  size: 'sm',
  spinner: false,
  type: 'button'
};

const Indicator = styled.span`
  align-items: center;
  display: ${({ isLoading }) => (isLoading ? 'flex' : 'none')};
  height: 100%;
  justify-content: center;
  left: 0;
  position: absolute;
  right: 0;
  text-align: center;
  top: 0;
  width: 100%;
`;

const Symbol = styled.span`
  visibility: ${({ isLoading }) => (isLoading ? 'hidden' : 'visible')};
`;

const Text = styled.span`
  margin-left: ${({ icon }) => (icon ? '7px' : '0')};
  visibility: ${({ isLoading }) => (isLoading ? 'hidden' : 'visible')};
`;

function Button({
  block,
  children,
  icon,
  isLoading,
  look,
  minimal,
  outline,
  passive,
  size,
  spinner,
  ...props
}) {
  const Tag = passive ? 'span' : 'button';
  let restProps = { ...props };

  // Filter unrelatd props
  if (passive) {
    restProps = omit(props, ['type']);
    restProps.role = 'button';
  }

  return (
    <Tag
      {...restProps}
      className={cx(
        'btn',
        block && 'btn-block',
        !!look && !outline && `btn-${look}`,
        `btn-${size}`,
        !!outline && `btn-outline-${outline}`,
        isLoading && 'disabled'
      )}
      disabled={isLoading}
    >
      {!!icon && (
        <Symbol isLoading={isLoading && spinner}>
          <Icon icon={icon} />
        </Symbol>
      )}
      {!!children && (
        <React.Fragment>
          <If condition={isBrowser || (!minimal && isMobile)}>
            <Text icon={icon} isLoading={isLoading && spinner}>
              {children}
            </Text>
          </If>
          <Indicator isLoading={isLoading && spinner}>
            <Spinner color="#fff" />
          </Indicator>
        </React.Fragment>
      )}
    </Tag>
  );
}

Button.propTypes = propTypes;
Button.defaultProps = defaultProps;

Button.Set = Set;

export default Button;
