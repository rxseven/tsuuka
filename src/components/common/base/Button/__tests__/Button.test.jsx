import React from 'react';
import { fireEvent } from 'react-testing-library';

import Icon from 'components/common/base/Icon';
import Spinner from 'components/common/base/Spinner';
import { factory, toMarkup } from 'tests/utilities';
import Button from '../index';

// Mock
jest.mock('components/common/base/Icon', () => jest.fn(() => <i />));
jest.mock('components/common/base/Spinner', () => jest.fn(() => <i />));

// Arrange
const context = expect.any(Object);
const seed = { content: 'content' };
const source = {
  content: seed.content,
  onClick: jest.fn()
};
const input = {
  ...seed,
  ...source,
  children: <span>{seed.content}</span>
};

// Setup
function setup(props) {
  return factory(Button, source, props);
}

// Test suites
describe('<Button />', () => {
  it('should render without crashing', () => {
    setup();
  });

  it('should render "button" tag with "button" type by default', () => {
    const expected = { type: 'button' };
    const { component } = setup();

    expect(component).toHaveAttribute('type', expected.type);
  });

  it('should render "button" tag with "submit" type', () => {
    const props = { type: 'submit' };
    const expected = { ...props };
    const { component } = setup(props);

    expect(component).toHaveAttribute('type', expected.type);
  });

  it('should render a passive button', () => {
    const props = { passive: true };
    const expected = { type: 'button' };
    const { component } = setup(props);

    expect(component).not.toHaveAttribute('type', expected.type);
  });

  it('should render a disabled button when loading data', () => {
    const props = { isLoading: true };
    const { component } = setup(props);

    expect(component).toBeDisabled();
  });

  it('should show a spinner when loading data', () => {
    const props = {
      children: input.children,
      isLoading: true,
      spinner: true
    };
    const expected = { called: { color: '#fff' } };
    const { getByTestId } = setup(props);

    expect(getByTestId('spinner')).toBeVisible();
    expect(Spinner).toHaveBeenCalledWith(expected.called, context);
  });

  it('should render a button as a "block" element', () => {
    const props = { block: true };
    const expected = { class: 'btn-block' };
    const { component } = setup(props);

    expect(component).toHaveClass(expected.class);
  });

  it('should render "danger" style', () => {
    const props = { look: 'danger' };
    const expected = { class: `btn-${props.look}` };
    const { component } = setup(props);

    expect(component).toHaveClass(expected.class);
  });

  it('should render "outline" style', () => {
    const props = { outline: 'primary' };
    const expected = { class: `btn-outline-${props.outline}` };
    const { component } = setup(props);

    expect(component).toHaveClass(expected.class);
  });

  it('should render "large" button', () => {
    const props = { size: 'lg' };
    const expected = { class: `btn-${props.size}` };
    const { component } = setup(props);

    expect(component).toHaveClass(expected.class);
  });

  it(`should contain "trash" icon`, () => {
    const props = { icon: 'trash' };
    const expected = { icon: { ...props } };
    setup(props);

    expect(Icon).toHaveBeenCalledWith(expected.icon, context);
  });

  it('should call onClick() when it is clicked', () => {
    const expected = { called: 1 };
    const { component } = setup();

    fireEvent.click(component);

    expect(input.onClick).toHaveBeenCalledTimes(expected.called);
  });

  it('should not render children by default', () => {
    const { component } = setup();

    expect(component).toBeEmpty();
  });

  it('should render passed children correctly', () => {
    const props = { children: input.children };
    const expected = {
      content: input.content,
      html: input.children
    };
    const { component } = setup(props);

    expect(component).toHaveTextContent(expected.content);
    expect(component).toContainHTML(toMarkup(expected.html));
  });
});
