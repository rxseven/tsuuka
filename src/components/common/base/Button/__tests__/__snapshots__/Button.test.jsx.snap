// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Snapshot tests should render correctly 1`] = `
<button
  className="btn btn-secondary btn-sm"
  disabled={false}
  onClick={[MockFunction]}
  type="button"
>
  <span
    className="sc-bxivhb IvOzD"
  >
    Text
  </span>
  <span
    className="sc-bwzfXH iDiJqN"
  >
    <span>
      <style>
         @-webkit-keyframes brsk-three-bounce { 0%, 80%, 100% { -webkit-transform: scale(0); } 40% { -webkit-transform: scale(1); } } @keyframes brsk-three-bounce { 0%, 80%, 100% { transform: scale(0); -webkit-transform: scale(0); } 40% { transform: scale(1); -webkit-transform: scale(1); } } 
      </style>
      <div
        style={
          Object {
            "WebkitAnimationDelay": "-0.32s",
            "WebkitAnimationDuration": "1.25s",
            "WebkitAnimationIterationCount": "infinite",
            "WebkitAnimationName": "brsk-three-bounce",
            "WebkitAnimationTimingFunction": "ease-in-out",
            "animationDelay": "-0.32s",
            "animationDuration": "1.25s",
            "animationIterationCount": "infinite",
            "animationName": "brsk-three-bounce",
            "animationTimingFunction": "ease-in-out",
            "backgroundColor": "#fff",
            "borderRadius": "100%",
            "display": "inline-block",
            "height": 7,
            "marginRight": 5,
            "width": 7,
          }
        }
      />
      <div
        style={
          Object {
            "WebkitAnimationDelay": "-0.16s",
            "WebkitAnimationDuration": "1.25s",
            "WebkitAnimationIterationCount": "infinite",
            "WebkitAnimationName": "brsk-three-bounce",
            "WebkitAnimationTimingFunction": "ease-in-out",
            "animationDelay": "-0.16s",
            "animationDuration": "1.25s",
            "animationIterationCount": "infinite",
            "animationName": "brsk-three-bounce",
            "animationTimingFunction": "ease-in-out",
            "backgroundColor": "#fff",
            "borderRadius": "100%",
            "display": "inline-block",
            "height": 7,
            "marginRight": 5,
            "width": 7,
          }
        }
      />
      <div
        style={
          Object {
            "WebkitAnimationDuration": "1.25s",
            "WebkitAnimationIterationCount": "infinite",
            "WebkitAnimationName": "brsk-three-bounce",
            "WebkitAnimationTimingFunction": "ease-in-out",
            "animationDuration": "1.25s",
            "animationIterationCount": "infinite",
            "animationName": "brsk-three-bounce",
            "animationTimingFunction": "ease-in-out",
            "backgroundColor": "#fff",
            "borderRadius": "100%",
            "display": "inline-block",
            "height": 7,
            "width": 7,
          }
        }
      />
    </span>
  </span>
</button>
`;
